Content.Server/StationEvents/Events/SecondLifeRule.cs

```diff
@@ -0,0 +1,181 @@
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<SecondLifeRuleComponent, TakeGhostRoleEvent>(OnTakeGhostRole);
+    }
+
+    private void RebuildRoles(EntityUid uid, SecondLifeRuleComponent component)
+    {
+        bool rolesDirty = false;
+        // Add and update roles
+        foreach (var roleKvp in component.Roles)
+        {
+            if (roleKvp.Value != 0)
+            {
+                // Check we haven't accepted all the entries for this role.
+                var numAccepted = component.AcceptedCount.GetValueOrDefault(roleKvp.Key, 0);
+                if (numAccepted >= roleKvp.Value)
+                {
+                    if (component.RoleInfo.TryGetValue(roleKvp.Key, out var ghostInfoDel))
+                    {
+                        // All the possible slots for this role have been used by players
+                        _ghostRole.UnregisterGhostRole(ghostInfoDel);
+                        component.RoleInfo.Remove(roleKvp.Key);
+                    }
+
+                    // Do not create or update this role, it's exhausted now.
+                    continue;
+                }
+            }
+
+            if (component.RoleInfo.TryGetValue(roleKvp.Key, out var ghostInfo))
+            {
+                // Already registered this role, just check it for updates.
+                if (component.Rules != ghostInfo.Rules || component.Description != ghostInfo.Description)
+                {
+                    ghostInfo.Rules = component.Rules;
+                    ghostInfo.Description = component.Description;
+
+                    // Only need to explicitly call UpdateAllEui if we only update rules and desc without
+                    //   adding or removing roles after.
+                    rolesDirty = true;
+                }
+            }
+            else
+            {
+                // This role requires registration. Perhaps the event just started or perhaps someone
+                //   added / increased limit on a role with VV
+
+                if (PrototypeManager.TryIndex<JobPrototype>(roleKvp.Key, out var job))
+                {
+                    var desc = (job.Description != null)
+                        ? $"{Loc.GetString(job.Description)} : {Loc.GetString(component.Description)}"
+                        : Loc.GetString(component.Description);
+
+                    var info = new GhostRoleInfo
+                    {
+                        Name = Loc.GetString(job.Name),
+                        Description = desc,
+                        Rules = Loc.GetString(component.Rules),
+                        Owner = uid, // All the Roles point back to the SecondLifeRuleComponent's entity.
+                        UserId = roleKvp.Key,
+                    };
+                    _ghostRole.RegisterGhostRole(ref info);
+                    component.RoleInfo[info.UserId] = info;
+                }
+            }
+        }
```

Comment:
am i understanding this code right that this system is bookkeeping what ghost roles should be open for itself (and only allowing 1 ghost role per job importantly)? is there any reason that it cant just create multiple ghost roles for each slot its opening? that would be in line with how all other ghost roles work

Reply:
If there's just one ghost role with multiple slots, It's easier to apply for. Otherwise its far more likely you'll get caught out racing for a specific role.

1. Click and wait for Passenger slot 1
2. Someone else takes it
3. Click and wait for Passenger slot 3
4. Someone else takes it

Wheras if the roles are shared, you will only miss out once all that type of role has been taken.